src/main/java/io/asyncer/r2dbc/mysql/QueryFlow.java

```diff
@@ -1027,13 +1060,31 @@ private HandshakeResponse createHandshakeResponse(int envelopeId, Capability cap
             authType = MySqlAuthProvider.CACHING_SHA2_PASSWORD;
         }
 
-        return HandshakeResponse.from(envelopeId, capability, context.getClientCollation().getId(),
-            user, authorization, authType, database, ATTRIBUTES);
+        return HandshakeResponse.from(capability, context.getClientCollation().getId(), user, authorization,
+            authType, database, ATTRIBUTES, zstdCompressionLevel);
+    }
+
+    private boolean isZstdAllowed(Capability capability) {
+        return capability.isZstdCompression() && compressions.contains(CompressionAlgorithm.ZSTD);
+    }
+
+    private boolean isZlibAllowed(Capability capability) {
+        return capability.isZlibCompression() && compressions.contains(CompressionAlgorithm.ZLIB);
     }
 
     private static String authFails(String authType, String phase) {
         return "Authentication type '" + authType + "' must require SSL in " + phase + " phase";
     }
+
+    private static boolean isZstdSupported() {
+        try {
+            Class.forName("com.github.luben.zstd.Zstd", false,
+                LoginExchangeable.class.getClassLoader());
```

Comment:
Could  consider use `AccessController.doPrivileged` to get classLoader? (incase of a SecurityManager)